go/src/koding/kites/kloud/provider/azure/azure.go

```diff
@@ -0,0 +1,207 @@
+)
+
+var p = &provider.Provider{
+	Name:         "azure",
+	ResourceName: "instance",
+	Machine:      newMachine,
+	Stack:        newStack,
+	Schema: &provider.Schema{
+		NewCredential: newCredential,
+		NewBootstrap:  newBootstrap,
+		NewMetadata:   newMetadata,
+	},
+}
+
+func init() {
+	provider.Register(p)
+}
+
+func newMachine(bm *provider.BaseMachine) (provider.Machine, error) {
+	m := &Machine{BaseMachine: bm}
+	cred := m.Cred()
+
+	c, err := management.ClientFromPublishSettingsDataWithConfig(cred.PublishSettings, cred.SubscriptionID, management.DefaultConfig())
+	if err != nil {
+		return nil, err
+	}
+
+	vmclient := vm.NewClient(c)
+
+	m.AzureClient = c
+	m.AzureVMClient = &vmclient
+
+	return m, nil
+}
+
+func newStack(bs *provider.BaseStack) (provider.Stack, error) {
+	return &Stack{BaseStack: bs}, nil
+}
+
+func newCredential() interface{} {
+	return &Cred{}
+}
+
+func newBootstrap() interface{} {
+	return &Bootstrap{}
+}
+
+func newMetadata(m *stack.Machine) interface{} {
+	if m == nil {
+		return &Meta{}
+	}
+
+	meta := &Meta{
+		InstanceID:      m.Attributes["id"],
+		HostedServiceID: m.Attributes["hosted_service_name"],
+		InstanceType:    m.Attributes["size"],
+	}
+
+	if cred, ok := m.Credential.Credential.(*Cred); ok {
+		meta.Location = cred.Location
+	}
+
+	return meta
+}
+
+type PublishData struct {
+	Subscriptions []Subscription `xml:"PublishProfile>Subscription"`
+}
+
+type Subscription struct {
+	ID string `xml:"Id,attr"`
+}
+
+// Cred represents jCredentialDatas.meta for "azure" provider.
+type Cred struct {
+	// Credentials.
+	PublishSettings []byte `json:"publish_settings" bson:"publish_settings" hcl:"publish_settings"` // required
+	SubscriptionID  string `json:"subscription_id,omitempty" bson:"subscription_id,omitempty" hcl:"subscription_id"`
+	Location        string `json:"location,omitempty" bson:"location,omitempty" hcl:"location"` // by default "East US 2"
+}
+
+var _ stack.Validator = (*Cred)(nil)
+
+func (meta *Cred) PublishData() (*PublishData, error) {
+	var pb PublishData
+
+	if err := xml.Unmarshal(meta.PublishSettings, &pb); err != nil {
+		return nil, err
+	}
+
+	if len(pb.Subscriptions) == 0 {
+		return nil, errors.New("no subscriptions found in the publish settings file")
+	}
+
+	return &pb, nil
+}
+
+// Valid implements the kloud.Validator interface.
+func (meta *Cred) Valid() error {
+	if meta.Location == "" {
+		meta.Location = "East US 2"
+	}
+	if len(meta.PublishSettings) == 0 {
+		return errors.New("publish settings are emtpty or missing")
+	}
+
+	pb, err := meta.PublishData()
+	if err != nil {
+		return err
+	}
+
+	// If SubscriptionID was not explicitely provided and the publish
+	// settings file contains only one subscription, we default to it.
+	if meta.SubscriptionID == "" {
+		if n := len(pb.Subscriptions); n != 1 {
+			return fmt.Errorf("publish settings contain %d subscriptions, please specify which one to use", n)
+		}
+
+		return nil
+	}
+
+	var found bool
```

Comment:
yup, will do - when I coded it originally I had two more checks after pb.Subscriptions that's why no early return here